Improves AI Law Module Examining

san7890 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

Closes https://github.com/tgstation/tgstation/issues/68604

## About The Pull Request

Hey there,

![image](https://user-images.githubusercontent.com/34697715/186795743-db5176cd-0471-4372-998b-cbb74c172fb3.png)

The old version of checking to see what AI laws were on a board were just ugly. They were in a proc completely separate from the actual Examine, and the to_chat utilization it used meant that ghosts couldn't see it (since they didn't count as mob/user i believe). So, let's use examine blocks, make the implementation cleaner in the examine (and actually make it ordered too!).

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

![image](https://user-images.githubusercontent.com/34697715/186795757-1e8ec73f-af72-4f49-a7eb-a6be1d43862a.png)

Does that look _so_ much better?

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
qol: When you examine or use an AI Law Module in-hand, it should look a lot nicer as you read the laws that you're about to subject that poor silicon to.
fix: When ghosts examine AI Law Modules, they are now able to see the actual laws on the board.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

san7890 committed:
Improves AI Law Module Examining

Hey there,

The old version of checking to see what AI laws were on a board were just ugly. They were in a proc completely separate from the actual Examine, and the to_chat utilization it used meant that ghosts couldn't see it (since they didn't count as mob/user i believe). So, let's use examine blocks, make the implementation cleaner in the examine (and actually make it ordered too!).

Mothblocks commented on code/game/objects/items/AI_modules/_AI_modules.dm:
```diff
@@ -30,22 +30,29 @@
+/// Show the laws for the module.
+/obj/item/ai_module/proc/display_laws()
+	/// Used to assemble the laws to show to an examining user.
+	var/assembled_laws
```
you dont need to doc comment this since it wont show anywhere

Mothblocks commented on code/game/objects/items/AI_modules/_AI_modules.dm:
```diff
@@ -30,22 +30,29 @@
 	return SHOULD_QDEL_MODULE //instead of the roundstart bid to un-unique the AI, there will be a research requirement for it.
 
-/obj/item/ai_module/proc/show_laws(mob/user as mob)
+/// Show the laws for the module.
```
```suggestion
/// Returns a text display of the laws for the module.
```

otherwise it implies side effects

Mothblocks commented on code/game/objects/items/AI_modules/_AI_modules.dm:
```diff
@@ -30,22 +30,29 @@
 /obj/item/ai_module/attack_self(mob/user as mob)
 	..()
-	show_laws(user)
+	var/examine_laws = display_laws()
+	to_chat(user, examine_block(examine_laws))
```
```suggestion
	to_chat(user, examine_block(display_laws()))
```

timothymtorres commented:
I know there was a problem with ghosts not seeing AI laws in #68604 recently that nobody was able to reproduce, it could be somehow related since the lawset weren't displaying.

Also `(mob/user as mob)` should apply to ghosts since they are apart of the `mob/dead` subtype.  (should be valid for any mob)

san7890 committed:
Addresses Review + File Organization

Cleans up comments, compresses lines of code.

I also moved the handle_unique_ai() proc below all the examine/attack/display_laws procs to give it some sense of grouping.

san7890 committed:
i accidentally deleted the whole line

san7890 commented:
> I know there was a problem with ghosts not seeing AI laws in #68604 recently that nobody was able to reproduce, it could be somehow related since the lawset wasn't displaying.

On my local, I noticed that if I joined in as an observer at the start of a shift and immediately went to examine a law board without ever spawning myself in as a mob, it wouldn't show the laws for whatever reason. Once I assumed a body, it would work fine. I'm not sure why it was breaking then, but we're using a new framework now that should not rely on the actual mob looking at it, so w/e.

timothymtorres commented:
> It actually could have been the `if(Adjacent(user))` check firing?

Bingo!  If you are a ghost you should be able to examine without being adjacent and was probably causing both issues.

timothymtorres commented on code/game/objects/items/AI_modules/_AI_modules.dm:
```diff
@@ -30,22 +30,28 @@
 
 /obj/item/ai_module/examine(mob/user as mob)
 	. = ..()
-	if(Adjacent(user))
```
I'd probably keep this in by doing:

```
if((isliving(user) && Adjacent(user)) || isobserver(user)
```

timothymtorres replied:
Also link this issue to #68604 to properly close it since the underlying issue was never resolved.

san7890 replied:
Isn't this just for the Module Board that contains the laws?

timothymtorres replied:
https://github.com/tgstation/tgstation/blob/fc91c2ed1f9e2311382472acb746feb69cdcca68/code/modules/mob/living/silicon/examine.dm#L1-L6

Damn, nevermind you are right.  I should have looked closer but I assumed the underlying issue was the same since it was both dealing with examine and laws.  

I'll reclose the issue.

Mothblocks commented on code/game/objects/items/AI_modules/_AI_modules.dm:
```diff
@@ -30,22 +30,28 @@
+/// Returns a text display of the laws for the module.
+/obj/item/ai_module/proc/display_laws()
+	// Used to assemble the laws to show to an examining user.
+	var/assembled_laws
```
```suggestion
	var/assembled_laws = ""
```

just to make sure we always return a string

san7890 committed:
Update code/game/objects/items/AI_modules/_AI_modules.dm

real

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>
